ops/model.py

```diff
@@ -230,6 +241,96 @@ def _get_unique(self, relation_name, relation_id=None):
+            name = binding_name.name
+            relation_id = binding_name.id
+        elif isinstance(binding_name, str):
+            name = key = binding_name
+            relation_id = None
+        else:
+            raise ModelError(f'binding_name must be a str or Relation, not {type(binding_name).__name__}')
+        binding = self._data.get(key)
+        if binding is None:
+            binding = Binding(name, relation_id, self._backend)
+            self._data[key] = binding
+        return binding
+
+
+class Binding:
+    """A representation of a binding to a space."""
+
+    def __init__(self, name, relation_id, backend):
+        self.name = name
+        self._relation_id = relation_id
+        self._backend = backend
+        self._network = None
+
+    @property
+    def network(self):
+        if self._network is None:
+            try:
+                if self._relation_id is None:
+                    network_info = self._backend.network_get(self.name)
+                else:
+                    network_info = self._backend.network_get(self.name, self._relation_id)
+            except RelationNotFoundError:
+                # If a relation is dead, we can still get network info associated with an endpoint itself
+                # not associated with a particular relation id.
+                network_info = self._backend.network_get(self.name)
+            self._network = Network(network_info['bind-addresses'], network_info['ingress-addresses'], network_info['egress-subnets'])
+        return self._network
+
+
+class Network:
+    """A representation of a unit's view of the network associated with a network space binding."""
+
+    def __init__(self, device_info, ingress_addresses, egress_subnets):
+        self.devices = []
+        for netdev in device_info:
+            self.devices.append(NetworkDevice(netdev['interface-name'], netdev['addresses']))
+
+        self.ingress_addresses = []
+        for address in ingress_addresses:
+            self.ingress_addresses.append(ipaddress.ip_address(address))
+        self.egress_subnets = []
+        for subnet in egress_subnets:
+            self.egress_subnets.append(ipaddress.ip_network(subnet))
+
+    @property
+    def bind_address(self):
+        return self.devices[0].addresses[0]
+
+    @property
+    def ingress_address(self):
+        return self.ingress_addresses[0]
+
+
+class NetworkDevice:
+
+    def __init__(self, name, address_info):
+        self.name = name
+        self.addresses = []
+        self.cidrs = []
+        for entry in address_info:
+            address = entry['value']
+            self.addresses.append(ipaddress.ip_address(address))
+            cidr = entry.get('cidr')
+            if cidr:
+                self.cidrs.append(ipaddress.ip_interface(cidr))
+            else:
+                # At least provide a /32 if there is no CIDR provided.
+                self.cidrs.append(ipaddress.ip_interface(entry['value']))
```

Comment:
Why is it not using ip_network here?  Having to index both address and cidrs by the same value to be able to associate one with the other feels a bit uncomfortable.

Reply:
`ip_network` only accepts network addresses, not host addresses:

```
ipaddress.ip_network('192.168.0.1/24')
# ...
ValueError: 192.168.0.1/24 has host bits set
```

And what I would like to see under `self.cidrs` is a list of things that can give me <host-address>/<prefix> if that information is available, not <network-address>/<prefix>. I am not sure why this wouldn't be available in general but Juju 2.7.2 doesn't return a CIDR with the localhost (LXD) provider so it's not reliably available.

Reply:
Addressed in https://github.com/canonical/operator/pull/109/commits/bb50c88674cccc35edfa70e80c2198b2187adc4e